File: effects/geometric.py
import cv2
import math
import random

from core.utils import calculate_2d_distance
from effects.particles import emit_particles

# ---------------------------------------------------------------------------
# Color Constants
# ---------------------------------------------------------------------------
COLOR_WHITE = (255, 255, 255)


# ---------------------------------------------------------------------------
# 3D Shape & Hologram Renderers
# ---------------------------------------------------------------------------

def draw_3d_cube(
    img,
    cx,
    cy,
    scale,
    angle,
    color
):
    """Renders rotating 3D wireframe cube."""

    rad = math.radians(angle)

    cos_a = math.cos(rad)
    sin_a = math.sin(rad)

    cos_b = math.cos(rad * 0.7)
    sin_b = math.sin(rad * 0.7)

    s = scale

    base_vertices = [
        [-s, -s, -s],
        [s, -s, -s],
        [s, s, -s],
        [-s, s, -s],
        [-s, -s, s],
        [s, -s, s],
        [s, s, s],
        [-s, s, s]
    ]

    screen_pts = []

    for x, y, z in base_vertices:

        xz_x = (
            x * cos_a
            + z * sin_a
        )

        xz_z = (
            -x * sin_a
            + z * cos_a
        )

        yz_y = (
            y * cos_b
            - xz_z * sin_b
        )

        px = int(cx + xz_x)
        py = int(cy + yz_y)

        screen_pts.append(
            (px, py)
        )

    edges = [
        (0, 1),
        (1, 2),
        (2, 3),
        (3, 0),
        (4, 5),
        (5, 6),
        (6, 7),
        (7, 4),
        (0, 4),
        (1, 5),
        (2, 6),
        (3, 7)
    ]

    for pt1, pt2 in edges:

        cv2.line(
            img,
            screen_pts[pt1],
            screen_pts[pt2],
            COLOR_WHITE,
            3,
            cv2.LINE_AA
        )

        cv2.line(
            img,
            screen_pts[pt1],
            screen_pts[pt2],
            color,
            1,
            cv2.LINE_AA
        )

    for pt in screen_pts:

        cv2.circle(
            img,
            pt,
            3,
            COLOR_WHITE,
            -1,
            cv2.LINE_AA
        )


def draw_cyber_shield(
    img,
    cx,
    cy,
    radius,
    angle,
    color
):
    """Renders futuristic rotating Cyber HUD Shield."""

    cv2.circle(
        img,
        (cx, cy),
        radius,
        color,
        2,
        cv2.LINE_AA
    )

    cv2.circle(
        img,
        (cx, cy),
        radius + 8,
        COLOR_WHITE,
        1,
        cv2.LINE_AA
    )

    cv2.circle(
        img,
        (cx, cy),
        int(radius * 0.6),
        color,
        1,
        cv2.LINE_AA
    )

    num_ticks = 12

    for i in range(num_ticks):

        a = math.radians(
            angle + i * (360 / num_ticks)
        )

        x1 = int(
            cx + (radius - 5) * math.cos(a)
        )

        y1 = int(
            cy + (radius - 5) * math.sin(a)
        )

        x2 = int(
            cx + (radius + 12) * math.cos(a)
        )

        y2 = int(
            cy + (radius + 12) * math.sin(a)
        )

        cv2.line(
            img,
            (x1, y1),
            (x2, y2),
            COLOR_WHITE if i % 3 == 0 else color,
            2,
            cv2.LINE_AA
        )

    l = 15

    cv2.line(
        img,
        (cx - l, cy),
        (cx + l, cy),
        color,
        1,
        cv2.LINE_AA
    )

    cv2.line(
        img,
        (cx, cy - l),
        (cx, cy + l),
        color,
        1,
        cv2.LINE_AA
    )


def draw_energy_core(
    img,
    cx,
    cy,
    scale,
    angle,
    color
):
    """Renders pulsing 3D energy core."""

    rad = math.radians(angle)

    for ring_idx in range(3):

        r_angle = (
            rad
            + ring_idx * (math.pi / 3)
        )

        axes = (
            scale,
            max(
                5,
                int(
                    scale
                    * abs(math.sin(r_angle))
                )
            )
        )

        cv2.ellipse(
            img,
            (cx, cy),
            axes,
            int(math.degrees(r_angle)),
            0,
            360,
            color,
            2,
            cv2.LINE_AA
        )

    cv2.circle(
        img,
        (cx, cy),
        int(scale * 0.4),
        COLOR_WHITE,
        -1,
        cv2.LINE_AA
    )

    cv2.circle(
        img,
        (cx, cy),
        int(scale * 0.5),
        color,
        2,
        cv2.LINE_AA
    )

    emit_particles(
        cx,
        cy,
        color,
        count=2
    )


def draw_pyramid_and_laser(
    img,
    tip_px,
    vector_dir,
    angle,
    color
):
    """Renders rotating 3D pyramid + laser."""

    cx, cy = tip_px

    rad = math.radians(angle)

    cos_a = math.cos(rad)
    sin_a = math.sin(rad)

    s = 35

    base_vertices = [
        [-s, s, -s],
        [s, s, -s],
        [s, s, s],
        [-s, s, s],
        [0, -s, 0]
    ]

    screen_pts = []

    for x, y, z in base_vertices:

        rx = (
            x * cos_a
            + z * sin_a
        )

        ry = y

        screen_pts.append(
            (
                int(cx + rx),
                int(cy + ry)
            )
        )

    edges = [
        (0, 1),
        (1, 2),
        (2, 3),
        (3, 0),
        (0, 4),
        (1, 4),
        (2, 4),
        (3, 4)
    ]

    for pt1, pt2 in edges:

        cv2.line(
            img,
            screen_pts[pt1],
            screen_pts[pt2],
            color,
            2,
            cv2.LINE_AA
        )

    apex_x, apex_y = screen_pts[4]

    laser_end = (
        apex_x,
        apex_y - 120
    )

    cv2.line(
        img,
        (apex_x, apex_y),
        laser_end,
        COLOR_WHITE,
        4,
        cv2.LINE_AA
    )

    cv2.line(
        img,
        (apex_x, apex_y),
        laser_end,
        color,
        2,
        cv2.LINE_AA
    )

    cv2.circle(
        img,
        laser_end,
        5,
        COLOR_WHITE,
        -1,
        cv2.LINE_AA
    )


def draw_3d_star(
    img,
    cx,
    cy,
    scale,
    angle,
    color
):
    """Renders rotating 3D octahedron star."""

    rad = math.radians(angle)

    cos_a = math.cos(rad)
    sin_a = math.sin(rad)

    s = scale

    verts = [
        [0, -s, 0],
        [0, s, 0],
        [-s, 0, 0],
        [s, 0, 0],
        [0, 0, -s],
        [0, 0, s]
    ]

    screen_pts = []

    for x, y, z in verts:

        rx = (
            x * cos_a
            + z * sin_a
        )

        ry = y

        screen_pts.append(
            (
                int(cx + rx),
                int(cy + ry)
            )
        )

    edges = [
        (0, 2),
        (0, 3),
        (0, 4),
        (0, 5),
        (1, 2),
        (1, 3),
        (1, 4),
        (1, 5),
        (2, 4),
        (4, 3),
        (3, 5),
        (5, 2)
    ]

    for pt1, pt2 in edges:

        cv2.line(
            img,
            screen_pts[pt1],
            screen_pts[pt2],
            color,
            2,
            cv2.LINE_AA
        )

    for pt in screen_pts:

        cv2.circle(
            img,
            pt,
            3,
            COLOR_WHITE,
            -1,
            cv2.LINE_AA
        )


def draw_electric_arcs(
    img,
    pt1,
    pt2,
    color
):
    """Renders dynamic plasma lightning."""

    dist = calculate_2d_distance(
        pt1,
        pt2
    )

    steps = max(
        5,
        int(dist / 15)
    )

    pts = [pt1]

    for i in range(1, steps):

        t = i / steps

        lx = int(
            pt1[0] * (1 - t)
            + pt2[0] * t
            + random.randint(-12, 12)
        )

        ly = int(
            pt1[1] * (1 - t)
            + pt2[1] * t
            + random.randint(-12, 12)
        )

        pts.append(
            (lx, ly)
        )

    pts.append(pt2)

    for i in range(len(pts) - 1):

        cv2.line(
            img,
            pts[i],
            pts[i + 1],
            COLOR_WHITE,
            3,
            cv2.LINE_AA
        )

        cv2.line(
            img,
            pts[i],
            pts[i + 1],
            color,
            1,
            cv2.LINE_AA
        )


def draw_status_badge(
    img,
    cx,
    cy,
    is_up,
    color
):
    """Renders holographic status badge."""

    w_box = 140
    h_box = 50

    x1 = cx - w_box // 2
    y1 = cy - h_box // 2

    x2 = cx + w_box // 2
    y2 = cy + h_box // 2

    overlay = img.copy()

    cv2.rectangle(
        overlay,
        (x1, y1),
        (x2, y2),
        (10, 10, 10),
        -1
    )

    cv2.addWeighted(
        overlay,
        0.6,
        img,
        0.4,
        0,
        img
    )

    cv2.rectangle(
        img,
        (x1, y1),
        (x2, y2),
        color,
        2,
        cv2.LINE_AA
    )

    label = (
        "LIKE [APPROVED]"
        if is_up
        else
        "DISLIKE [REJECTED]"
    )

    arrow = "^" if is_up else "v"

    cv2.putText(
        img,
        label,
        (x1 + 10, cy + 5),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.45,
        color,
        2
    )

    cv2.putText(
        img,
        arrow,
        (x2 - 20, cy + 6),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.7,
        COLOR_WHITE,
        2
    )

File: website/effects/geometric.js
// ============================================================
// Geometric World Renderers (Ported from effects/geometric.py)
// ============================================================

import { getLandmarkPx } from "../tracking/gestureDetector.js";

const HAND_CONNECTIONS = [
    [0, 1], [1, 2], [2, 3], [3, 4],         // Thumb
    [0, 5], [5, 6], [6, 7], [7, 8],         // Index
    [5, 9], [9, 10], [10, 11], [11, 12],    // Middle
    [9, 13], [13, 14], [14, 15], [15, 16],  // Ring
    [13, 17], [17, 18], [18, 19], [19, 20], // Pinky
    [0, 17]                                  // Palm base
];

export function drawHandSkeleton(ctx, landmarks, primaryColor, secondaryColor, w, h) {
    const pts = landmarks.map(lm => getLandmarkPx(lm, w, h));

    ctx.save();
    ctx.lineWidth = 2;
    ctx.strokeStyle = primaryColor;
    HAND_CONNECTIONS.forEach(([p1Idx, p2Idx]) => {
        ctx.beginPath();
        ctx.moveTo(pts[p1Idx].x, pts[p1Idx].y);
        ctx.lineTo(pts[p2Idx].x, pts[p2Idx].y);
        ctx.stroke();
    });

    pts.forEach((pt, i) => {
        if ([4, 8, 12, 16, 20].includes(i)) {
            ctx.beginPath();
            ctx.arc(pt.x, pt.y, 6, 0, Math.PI * 2);
            ctx.fillStyle = secondaryColor;
            ctx.fill();

            ctx.beginPath();
            ctx.arc(pt.x, pt.y, 9, 0, Math.PI * 2);
            ctx.strokeStyle = primaryColor;
            ctx.lineWidth = 1.5;
            ctx.stroke();
        } else {
            ctx.beginPath();
            ctx.arc(pt.x, pt.y, 3, 0, Math.PI * 2);
            ctx.fillStyle = "#282828";
            ctx.fill();
            ctx.strokeStyle = primaryColor;
            ctx.lineWidth = 1;
            ctx.stroke();
        }
    });
    ctx.restore();
}

export function draw3DCube(ctx, cx, cy, scale, angle, color = "#00f0ff") {
    const rad = angle * Math.PI / 180;
    const cosA = Math.cos(rad);
    const sinA = Math.sin(rad);
    const cosB = Math.cos(rad * 0.7);
    const sinB = Math.sin(rad * 0.7);

    const s = scale;
    const baseVertices = [
        [-s, -s, -s], [s, -s, -s], [s, s, -s], [-s, s, -s],
        [-s, -s, s],  [s, -s, s],  [s, s, s],  [-s, s, s]
    ];

    const screenPts = baseVertices.map(([x, y, z]) => {
        const xzX = x * cosA + z * sinA;
        const xzZ = -x * sinA + z * cosA;
        const yzY = y * cosB - xzZ * sinB;
        return { x: cx + xzX, y: cy + yzY };
    });

    const edges = [
        [0, 1], [1, 2], [2, 3], [3, 0],
        [4, 5], [5, 6], [6, 7], [7, 4],
        [0, 4], [1, 5], [2, 6], [3, 7]
    ];

    ctx.save();
    ctx.lineCap = "round";

    ctx.lineWidth = 4;
    ctx.strokeStyle = "#ffffff";
    edges.forEach(([a, b]) => {
        ctx.beginPath();
        ctx.moveTo(screenPts[a].x, screenPts[a].y);
        ctx.lineTo(screenPts[b].x, screenPts[b].y);
        ctx.stroke();
    });

    ctx.lineWidth = 2;
    ctx.strokeStyle = color;
    edges.forEach(([a, b]) => {
        ctx.beginPath();
        ctx.moveTo(screenPts[a].x, screenPts[a].y);
        ctx.lineTo(screenPts[b].x, screenPts[b].y);
        ctx.stroke();
    });

    screenPts.forEach((pt) => {
        ctx.beginPath();
        ctx.arc(pt.x, pt.y, 3, 0, Math.PI * 2);
        ctx.fillStyle = "#ffffff";
        ctx.fill();
    });
    ctx.restore();
}

export function drawCyberShield(ctx, cx, cy, radius, angle, color = "#00f0ff") {
    ctx.save();

    ctx.beginPath();
    ctx.arc(cx, cy, radius, 0, Math.PI * 2);
    ctx.strokeStyle = color;
    ctx.lineWidth = 2.5;
    ctx.stroke();

    ctx.beginPath();
    ctx.arc(cx, cy, radius + 8, 0, Math.PI * 2);
    ctx.strokeStyle = "#ffffff";
    ctx.lineWidth = 1;
    ctx.stroke();

    ctx.beginPath();
    ctx.arc(cx, cy, radius * 0.6, 0, Math.PI * 2);
    ctx.strokeStyle = color;
    ctx.lineWidth = 1.5;
    ctx.stroke();

    const numTicks = 12;
    for (let i = 0; i < numTicks; i++) {
        const a = (angle + i * (360 / numTicks)) * Math.PI / 180;
        const x1 = cx + (radius - 5) * Math.cos(a);
        const y1 = cy + (radius - 5) * Math.sin(a);
        const x2 = cx + (radius + 12) * Math.cos(a);
        const y2 = cy + (radius + 12) * Math.sin(a);

        ctx.beginPath();
        ctx.moveTo(x1, y1);
        ctx.lineTo(x2, y2);
        ctx.strokeStyle = i % 3 === 0 ? "#ffffff" : color;
        ctx.lineWidth = 2;
        ctx.stroke();
    }

    const l = 15;
    ctx.beginPath();
    ctx.moveTo(cx - l, cy);
    ctx.lineTo(cx + l, cy);
    ctx.moveTo(cx, cy - l);
    ctx.lineTo(cx, cy + l);
    ctx.strokeStyle = color;
    ctx.lineWidth = 1.5;
    ctx.stroke();

    ctx.restore();
}

export function drawEnergyCore(ctx, cx, cy, scale, angle, color = "#ff8c00", particleSystem = null) {
    ctx.save();
    const rad = angle * Math.PI / 180;

    for (let ringIdx = 0; ringIdx < 3; ringIdx++) {
        const rAngle = rad + ringIdx * (Math.PI / 3);
        const rx = scale;
        const ry = Math.max(5, scale * Math.abs(Math.sin(rAngle)));

        ctx.save();
        ctx.translate(cx, cy);
        ctx.rotate(rAngle);
        ctx.beginPath();
        ctx.ellipse(0, 0, rx, ry, 0, 0, Math.PI * 2);
        ctx.strokeStyle = color;
        ctx.lineWidth = 2;
        ctx.stroke();
        ctx.restore();
    }

    ctx.beginPath();
    ctx.arc(cx, cy, scale * 0.4, 0, Math.PI * 2);
    ctx.fillStyle = "#ffffff";
    ctx.fill();

    ctx.beginPath();
    ctx.arc(cx, cy, scale * 0.5, 0, Math.PI * 2);
    ctx.strokeStyle = color;
    ctx.lineWidth = 2;
    ctx.stroke();

    ctx.restore();

    if (particleSystem) {
        particleSystem.emit(cx, cy, color, 2);
    }
}

export function drawPyramidAndLaser(ctx, tipPx, angle, color = "#ffee00") {
    const cx = tipPx.x;
    const cy = tipPx.y;
    const rad = angle * Math.PI / 180;
    const cosA = Math.cos(rad);
    const sinA = Math.sin(rad);

    const s = 35;
    const baseVertices = [
        [-s, s, -s],
        [s, s, -s],
        [s, s, s],
        [-s, s, s],
        [0, -s, 0]
    ];

    const screenPts = baseVertices.map(([x, y, z]) => {
        const rx = x * cosA + z * sinA;
        const ry = y;
        return { x: cx + rx, y: cy + ry };
    });

    const edges = [
        [0, 1], [1, 2], [2, 3], [3, 0],
        [0, 4], [1, 4], [2, 4], [3, 4]
    ];

    ctx.save();
    ctx.strokeStyle = color;
    ctx.lineWidth = 2;
    edges.forEach(([a, b]) => {
        ctx.beginPath();
        ctx.moveTo(screenPts[a].x, screenPts[a].y);
        ctx.lineTo(screenPts[b].x, screenPts[b].y);
        ctx.stroke();
    });

    const apex = screenPts[4];
    const laserEnd = { x: apex.x, y: apex.y - 120 };

    ctx.beginPath();
    ctx.moveTo(apex.x, apex.y);
    ctx.lineTo(laserEnd.x, laserEnd.y);
    ctx.strokeStyle = "#ffffff";
    ctx.lineWidth = 4;
    ctx.stroke();

    ctx.beginPath();
    ctx.moveTo(apex.x, apex.y);
    ctx.lineTo(laserEnd.x, laserEnd.y);
    ctx.strokeStyle = color;
    ctx.lineWidth = 2;
    ctx.stroke();

    ctx.beginPath();
    ctx.arc(laserEnd.x, laserEnd.y, 5, 0, Math.PI * 2);
    ctx.fillStyle = "#ffffff";
    ctx.fill();

    ctx.restore();
}

export function draw3DStar(ctx, cx, cy, scale, angle, color = "#00ff66") {
    const rad = angle * Math.PI / 180;
    const cosA = Math.cos(rad);
    const sinA = Math.sin(rad);

    const s = scale;
    const verts = [
        [0, -s, 0],
        [0, s, 0],
        [-s, 0, 0],
        [s, 0, 0],
        [0, 0, -s],
        [0, 0, s]
    ];

    const screenPts = verts.map(([x, y, z]) => {
        const rx = x * cosA + z * sinA;
        const ry = y;
        return { x: cx + rx, y: cy + ry };
    });

    const edges = [
        [0, 2], [0, 3], [0, 4], [0, 5],
        [1, 2], [1, 3], [1, 4], [1, 5],
        [2, 4], [4, 3], [3, 5], [5, 2]
    ];

    ctx.save();
    ctx.strokeStyle = color;
    ctx.lineWidth = 2;
    edges.forEach(([a, b]) => {
        ctx.beginPath();
        ctx.moveTo(screenPts[a].x, screenPts[a].y);
        ctx.lineTo(screenPts[b].x, screenPts[b].y);
        ctx.stroke();
    });

    screenPts.forEach(pt => {
        ctx.beginPath();
        ctx.arc(pt.x, pt.y, 3, 0, Math.PI * 2);
        ctx.fillStyle = "#ffffff";
        ctx.fill();
    });

    ctx.restore();
}

export function drawElectricArcs(ctx, pt1, pt2, color = "#ff00a0") {
    const dist = Math.hypot(pt1.x - pt2.x, pt1.y - pt2.y);
    const steps = Math.max(5, Math.floor(dist / 15));

    const pts = [pt1];
    for (let i = 1; i < steps; i++) {
        const t = i / steps;
        const lx = pt1.x * (1 - t) + pt2.x * t + (Math.random() * 24 - 12);
        const ly = pt1.y * (1 - t) + pt2.y * t + (Math.random() * 24 - 12);
        pts.push({ x: lx, y: ly });
    }
    pts.push(pt2);

    ctx.save();
    for (let i = 0; i < pts.length - 1; i++) {
        ctx.beginPath();
        ctx.moveTo(pts[i].x, pts[i].y);
        ctx.lineTo(pts[i + 1].x, pts[i + 1].y);
        ctx.strokeStyle = "#ffffff";
        ctx.lineWidth = 3;
        ctx.stroke();

        ctx.beginPath();
        ctx.moveTo(pts[i].x, pts[i].y);
        ctx.lineTo(pts[i + 1].x, pts[i + 1].y);
        ctx.strokeStyle = color;
        ctx.lineWidth = 1.5;
        ctx.stroke();
    }
    ctx.restore();
}

export function drawStatusBadge(ctx, cx, cy, isUp, color = "#00ff66") {
    const wBox = 140;
    const hBox = 50;
    const x1 = cx - wBox / 2;
    const y1 = cy - hBox / 2;

    ctx.save();
    ctx.fillStyle = "rgba(10, 15, 22, 0.85)";
    ctx.fillRect(x1, y1, wBox, hBox);

    ctx.strokeStyle = color;
    ctx.lineWidth = 2;
    ctx.strokeRect(x1, y1, wBox, hBox);

    const label = isUp ? "LIKE [APPROVED]" : "DISLIKE [REJECTED]";
    const arrow = isUp ? "▲" : "▼";

    ctx.font = "bold 11px system-ui, sans-serif";
    ctx.fillStyle = color;
    ctx.fillText(label, x1 + 8, cy + 28);

    ctx.font = "bold 14px system-ui, sans-serif";
    ctx.fillStyle = "#ffffff";
    ctx.fillText(arrow, x1 + wBox - 20, cy + 29);

    ctx.restore();
}

export function drawDualPinchCage(ctx, ptLeft, ptRight, angle, color = "#00ff66") {
    const cx = (ptLeft.x + ptRight.x) / 2;
    const cy = (ptLeft.y + ptRight.y) / 2;
    const dist = Math.hypot(ptLeft.x - ptRight.x, ptLeft.y - ptRight.y);

    ctx.save();
    ctx.beginPath();
    ctx.moveTo(ptLeft.x, ptLeft.y);
    ctx.lineTo(ptRight.x, ptRight.y);
    ctx.strokeStyle = "#ffffff";
    ctx.lineWidth = 3;
    ctx.stroke();

    ctx.beginPath();
    ctx.moveTo(ptLeft.x, ptLeft.y);
    ctx.lineTo(ptRight.x, ptRight.y);
    ctx.strokeStyle = color;
    ctx.lineWidth = 1.5;
    ctx.stroke();

    draw3DCube(ctx, cx, cy, Math.max(25, dist * 0.35), angle, color);

    [ptLeft, ptRight].forEach(pt => {
        ctx.beginPath();
        ctx.arc(pt.x, pt.y, 12, 0, Math.PI * 2);
        ctx.strokeStyle = "#ffffff";
        ctx.lineWidth = 2;
        ctx.stroke();
    });

    ctx.restore();
}
